base.go

```diff
@@ -115,6 +115,12 @@ func (db *baseDB) initConn(ctx context.Context, cn *pool.Conn) error {
 		}
 	}
 
+	if db.opt.BeforeConnect != nil {
+		if err := db.opt.BeforeConnect(ctx, db.opt); err != nil {
```

Comment:
I believe we should copy the options to avoid races:

```
opt := db.opt.Clone()
// use the cloned opt
```

Reply:
and then reassign to db.opt?

```
if db.opt.BeforeConnect != nil {
    opts := db.copt.Clone()
    if err := opt.BeforeConnect(ctx, opt); err != nil { return err }
    db.opts = opts		
}
```

Reply:
If you re assign to db.opt then you need to make that re assignment atomic. What if this code is being executed in multiple go routines?

Reply:
Also if we replace the db.opt here this can lead to a few interesting issues depending on how the user is using the library. What if they change the DB address entirely? what if they add TLS? Beyond simply changing the password or the user that a connection is issued with; changing other variables available in the options object could have a lot of unintended consequences that are currently unknown.

If at all possible we might just want to limit being able to change authentication parameters if we do update the original `db.opts`. Maybe limiting it to changing the user, password and TLS config. 

If they for some reason changed something like the connection pool settings, how would that affect the pool that is already running?

Reply:
@justcompile I don't think you need to replace `db.opt`. Just clone it (when needed for BeforeConnect) and use below in startup:

```go
opt := db.opt.Clone()

....

db.startup(ctx, cn, opt.User, opt.Password, opt.Database, opt.ApplicationName)

```

That will not cover all options but is good enough for this particular use case. We can improve it later if needed.

Reply:
```
if db.opt.BeforeConnect != nil {
		db.opt.mux.Lock()
		updateOpts := &BeforeConnectOptions{
			User: db.opt.User,
			Password: db.opt.Password,
			TLSConfig: db.opt.TLSConfig,
		}
		if err := db.opt.BeforeConnect(ctx, updateOpts); err != nil {
			db.opt.mux.Unlock()
			return err
		}
		
		db.opt.applyUpdatedConnectionOptions(updateOpts)
		db.opt.mux.Unlock()
	}
	
// options.go
func (opt *Options) applyUpdatedConnectionOptions(bcOpts *BeforeConnectOptions) {
	opt.User = bcOpts.User
	opt.Password = bcOpts.Password

	if bcOpts.TLSConfig != nil {
		opt.TLSConfig = bcOpts.TLSConfig.Clone()
	}
}
```

Based on the concurrency issue mentioned & the points around unintended side effects of a user changing the Addr or Pool options, I was looking at something like the above.

But if that's overkill, I can go with the opts.Clone approach

Reply:
Let's keep as is `BeforeConnect(ctx context.Context, o *pg.Options) error`. It will not work for all options but that is fine. It is an advanced feature. Add add a warning to the comment.